File: LeetCode-Swift/LeetCode-Swift/二叉树/_99_恢复二叉搜索树.swift
//
//  _99_恢复二叉搜索树.swift
//  LeetCode-Swift
//
//  Created by warden on 2020/2/17.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/recover-binary-search-tree/

import Foundation

class _99_恢复二叉搜索树 {
    func recoverTree(_ root: TreeNode?) {
        var first: TreeNode?
        var last: TreeNode?
        var prev: TreeNode?
        var node = root
        while (node != nil) {
            var tmp = node?.left
            while (tmp != nil && tmp!.right != nil && tmp?.right !== node) {
                tmp = tmp!.right
            }
            if (tmp == nil || tmp?.right === node) {
                if (tmp != nil) {
                    tmp?.right = nil
                }
                if (prev != nil && node!.val < prev!.val) {
                    last = node
                    first = first == nil ? prev : first
                }
                prev = node
                node = node?.right
            } else {
                tmp?.right = node
                node = node?.left
            }
        }
        if first != nil && last != nil {
            let tmp = last!.val
            last!.val = first!.val
            first!.val = tmp
        }
    }
}

/**
 输入: [1,3,null,null,2]

    1
   /
  3
   \
    2

 输出: [3,1,null,null,2]

    3
   /
  1
   \
    2
 示例 2:

 输入: [3,1,4,null,null,2]

   3
  / \
 1   4
    /
   2

 输出: [2,1,4,null,null,3]

   2
  / \
 1   4
    /
   3
 */
func test_99() {
    let solution = _99_恢复二叉搜索树();
    var root = generateTreeByLevel([1,3,nil,nil,2])
//    solution.recoverTree(root)
//    printTreeByLevel(root)
    root = generateTreeByLevel([3,1,4,nil,nil,2])
    solution.recoverTree(root)
    printTreeByLevel(root)
}

File: LeetCode-C++/LeetCode-C++/二叉树/_99_恢复二叉搜索树.cpp
//
//  _99_恢复二叉搜索树.cpp
//  LeetCode-C++
//
//  Created by warden on 2020/2/17.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/recover-binary-search-tree/

#include <stdio.h>
#include <algorithm>
#include <vector>
#include "TreeNode.hpp"

using namespace std;

class Solution {
public:
    void recoverTree(TreeNode* root) {
        TreeNode* first = NULL;
        TreeNode* last = NULL;
        TreeNode* prev = NULL;
        
        while (root) {
            TreeNode* tmp=root->left;
            while (tmp && (tmp->right && tmp->right != root))
                tmp = tmp->right;
            if (!tmp || tmp->right == root) {
                if (tmp)
                    tmp->right = NULL;
                if (prev && root->val < prev->val) {
                    last = root;
                    first = first == NULL ? prev : first;
                }
                prev = root;
                root = root->right;
            } else {
                tmp->right = root;
                root = root->left;
            }
        }
        if (first && last)
            std::swap(first->val, last->val);
    }
};

/**
 输入: [1,3,null,null,2]

    1
   /
  3
   \
    2

 输出: [3,1,null,null,2]

    3
   /
  1
   \
    2
 示例 2:

 输入: [3,1,4,null,null,2]

   3
  / \
 1   4
    /
   2

 输出: [2,1,4,null,null,3]

   2
  / \
 1   4
    /
   3
 */
/**
int main(int argc, const char * argv[]) {
    Solution* solution = new Solution();
    vector<int> nodes = {1,3,INT_MAX,INT_MAX,2};
    TreeNode* root = generateTreeByLevel(nodes);
//    solution.recoverTree(root);
    nodes = {3,1,4,INT_MAX,INT_MAX,2};
    root = generateTreeByLevel(nodes);
    solution->recoverTree(root);
    printTreeByLevel(root);
    return 0;
}
*/
